Implement queue_event move assignment via static move/dtor_impl

palarson opened the pull request:
Addresses issue described in https://github.com/boost-experimental/sml/issues/279

krzysztof-jusiak commented on include/boost/sml.hpp:
```diff
@@ -516,7 +516,15 @@ class queue_event {
     move(data, static_cast<queue_event &&>(other));
   }
-  queue_event &operator=(queue_event &&) = default;
+  queue_event &operator=(queue_event &&other) {
```
Thanks, @palarson, would you be able to provide test to cover the issue and keep the code working in the future 👍

palarson commented:
Added a regression test for the issue I saw - it's a little verbose because the issue only manifested when the defer event was being destroyed from the back half of the deque.

Verified that the test fails in master, and passes after this fix:

Before fix:

Test project /Users/plarson/upstream-sml/sml/BUILD
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 20
    Start 20: test_actions_defer

20: Test command: /Users/plarson/upstream-sml/sml/BUILD/test/ft/test_actions_defer
20: Test timeout computed to be: 1500
20: /Users/plarson/upstream-sml/sml/test/ft/actions_defer.cpp:289:e1data3.use_count() == 1
1/1 Test #20: test_actions_defer ...............***Failed    0.00 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.01 sec

The following tests FAILED:
   20 - test_actions_defer (Failed)
Errors while running CTest



After fix:
Test project /Users/plarson/upstream-sml/sml/BUILD
Constructing a list of tests
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 20
    Start 20: test_actions_defer

20: Test command: /Users/plarson/upstream-sml/sml/BUILD/test/ft/test_actions_defer
20: Test timeout computed to be: 1500
1/1 Test #20: test_actions_defer ...............   Passed    0.01 sec

The following tests passed:
  test_actions_defer

100% tests passed, 0 tests failed out of 1
